src/main/flight/alt_hold.c

```diff
@@ -90,65 +89,67 @@ void altHoldProcessTransitions(void) {
-    // - fine-tuning height is possible, eg if there is slow sensor drift
-    // - to keep the craft stable, throttle must be in the deadband, making exits smoother
-
-    const float rcThrottle = rcCommand[THROTTLE];
-
-    const float lowThreshold = 0.5f * (autopilotConfig()->hover_throttle + PWM_RANGE_MIN); // halfway between hover and MIN, e.g. 1150 if hover is 1300
-    const float highThreshold = 0.5f * (autopilotConfig()->hover_throttle + PWM_RANGE_MAX); // halfway between hover and MAX, e.g. 1650 if hover is 1300
-
-    float throttleAdjustmentFactor = 0.0f;
-    if (rcThrottle < lowThreshold) {
-        throttleAdjustmentFactor = scaleRangef(rcThrottle, PWM_RANGE_MIN, lowThreshold, -1.0f, 0.0f);
-    } else if (rcThrottle > highThreshold) {
-        throttleAdjustmentFactor = scaleRangef(rcThrottle, highThreshold, PWM_RANGE_MAX, 0.0f, 1.0f);
+    // User can adjust the target altitude with throttle, but only when
+    // - throttle is outside deadband, and 
+    // - throttle is not low (zero), and 
+    // - deadband is not configured to zero
+
+    float stickFactor = 0.0f;
+
+    if (altHold.allowStickAdjustment && calculateThrottleStatus() != THROTTLE_LOW) {
+        const float rcThrottle = rcCommand[THROTTLE];
+        const float lowThreshold = apConfig()->hover_throttle - altHold.deadband * (apConfig()->hover_throttle - PWM_RANGE_MIN);
+        const float highThreshold = apConfig()->hover_throttle + altHold.deadband * (PWM_RANGE_MAX - apConfig()->hover_throttle);
+
+        if (rcThrottle < lowThreshold) {
+            stickFactor = scaleRangef(rcThrottle, PWM_RANGE_MIN, lowThreshold, -1.0f, 0.0f);
+        } else if (rcThrottle > highThreshold) {
+            stickFactor = scaleRangef(rcThrottle, highThreshold, PWM_RANGE_MAX, 0.0f, 1.0f);
+        }
     }
-    // if failsafe is active, and we get here, we are in failsafe landing mode
+
+    // if failsafe is active, and we get here, we are in failsafe landing mode, it controls throttle
     if (failsafeIsActive()) {
         // descend at up to 10 times faster when high
-        // default landing time is now 60s; need to get the quad down in this time from reasonable height
-        // need a rapid descent if high to avoid that timeout, and must slow down closer to ground
+        // default landing timeout is now 60s; must to get the quad down within this limit
+        // need a rapid descent when initiated high, and must slow down closer to ground
         // this code doubles descent rate at 20m, to max 10x (10m/s on defaults) at 200m
-        // user should be able to descend within 60s from around 150m high without disarming, on defaults
         // the deceleration may be a bit rocky if it starts very high up
         // constant (set) deceleration target in the last 2m
-        throttleAdjustmentFactor = -(0.9f + constrainf(getAltitudeCm() / 2000.0f, 0.1f, 9.0f));
+        stickFactor = -(0.9f + constrainf(getAltitudeCm() / 2000.0f, 0.1f, 9.0f));
```

Comment:
Velocity control while the sticks are outside deadband is exactly what we have, though it was a bit of a 'kludge' on my part.
The 'velocity mode' just involves moving the target altitude at a rate equal to the intended velocity, and letting the PIDs do their thing. There was some basic iTerm suppression and some direct stick feedforward, that was about all, and the results were very good.  Later we can look at things that may improve performance, for example the iTerm suppression is a bit basic, when sticks move, but seems quite adequate for the relatively slow climb and descend rates being commanded; it does a reasonable, but not excellent, job of arresting extreme climbs or drops.
The intended benefit from this approach was that it avoided the need for any specific 'velocity' controller for altitude; the same controller could run all the time, whether or not the target vertical velocity was zero or some other value.
I guess the controller is a full-time velocity controller in a sense; when sticks are within deadband, target velocity is zero.  It's almost an academic/technical theoretical question as to whether it is a velocity controller or a position controller when the target velocity is zero.

Reply:
It is usually easier to decompose problems into smaller parts - in this case cascaded position-velocity controller.
It is possible to get it working using ' iTerm suppression', 'direct stick feedforward', 'stickFactor' etc. But quite often you end up wil complicated, unreadables 'full magic' controller, with performance similar to classic, simple design.

Reinventing two centuries of control theory development has its limitations ...

Reply:
In practice, it is not possible to do get good responses without some feedforward and without suppression of iTerm in situations where it would otherwise grow inappropriately.

The reason racers around the world use betaflight is because these customisations provide the 'magic' that makes it fly more precisely and accurately than anything else.

For this function I carefully tested whether or not the additional stuff helps, and yes, it makes a huge difference.  You can test your self by setting feedforward for altitude to zero, and by wrecking iTerm relax in a custom build :-)

I think if we ever proposed a generic controller we would need to include iTerm relax, D boosting and feedforward as standard parameters.  They are absolutely necessary if you want tight control.

Reply:
More generally I don't really understand the need for separate velocity and position controllers.
Velocity is just the derivative of position, so the PID parameters are and numerically interchangeable:
- position I == second integral of velocity (useless for velocity control)
- position P == velocity I
- position D == velocity
- position A == velocity D

Reply:
> The reason racers around the world use betaflight is because these customisations provide the 'magic' that makes it fly more precisely and accurately than anything else.

There are similar statements in audiophile group ;-)

> For this function I carefully tested whether or not the additional stuff helps, and yes, it makes a huge difference. You can test your self by setting feedforward for altitude to zero, and by wrecking iTerm relax in a custom build :-)

But you have chosen one semi-fixed structure and empirically tuned the parameters. All that is _can_ signify is that you are better at tuning system with additional tunable parameters. 
Some term of I-term windup handling is necessary; there are some interesting methods in control theory toolbox that would probably easily outperform 'relax'  method.
Not sure about feedforward, in some structures it can be factored into other coefficients.
D boosting .. hard to tell without some simulation. At least some parts of BF use it to trade stability for speed. 

> Velocity is just the derivative of position, so the PID parameters are and numerically interchangeable:

Mathematically, it is equivalent or almost equivalent. But you can also construct such PID as IIR filter, and tune that filter response.

With cascaded structure:you have:
- better control over speed limiting (this PR does arbitrary limit on DA instead)
- easier switching between pos and vel mode (i-term disabling here)
- easier/no setpoint scheduling (complicated stepping of position; initially slow braking)
- possibly easier tuning (tune 2-parameter velocity PD first, than tune simple PI)

Reply:
This is my first attempt at an altitude control concept, and I’m sure it can be improved. You’re welcome to try :-)

I have improved the behaviour a bit in the next PR and I’m sure you will like it not one bit. But it works well. 

All theory has to bow to algorithms that actually work. The code that write works and provides some tools for users to optimise the performance. 

Other people can - and should - propose changes or improvements. The proposer must fly their code, make logs and show the data that indicates that it does actually perform better. If such a change is shown to improve performance without excessively complicating the code then of course we would welcome it. 

Many people have suggested that including accelerometer data would improve altitude control or position hold. Since altitude hold was released no-one has provided code that has achieved this. Perhaps once this PR is released someone may look into it.

Reply:
D boosting is what lets the quad stop fast from high speed, because a lot of pre-existing D means an immediately strong angle for the stop. But the right amount of D for high speed stops is too much for a stable hover, it’s too reactive. I know because I tested these things very carefully. We can probably do it better with more experience and a careful evaluation of the functional strengths and weaknesses of the current code.

Reply:
> All theory has to bow to algorithms that actually work. 

I'm so glad that BF is not written in perl ...

Reply:
IDK... but fore sure, I'm glad that I can make it fly, and that you know how to make great code :-)